src/bin/pg_autoctl/monitor.c

```diff
@@ -1190,3 +1199,261 @@ monitor_get_notifications(Monitor *monitor)
+{
+	PGSQL *pgsql = &monitor->pgsql;
+
+	return pgsql_alter_extension_update_to(pgsql,
+										   PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+										   targetVersion);
+}
+
+
+/*
+ * monitor_ensure_extension_version checks that we are running a extension
+ * version on the monitor that we are compatible with in pg_autoctl. If that's
+ * not the case, we blindly try to update the extension version on the monitor
+ * to the target version we have in our default.h.
+ *
+ * NOTE: we don't check here if that's an upgrade or a downgrade, we rely on
+ * the extension's upgrade path to be free of downgrade paths (such as
+ * pgautofailover--1.2--1.1.sql).
+ */
+bool
+monitor_ensure_extension_version(Monitor *monitor)
+{
+	MonitorExtensionVersion version = { 0 };
+
+	if (!monitor_get_extension_version(monitor, &version))
+	{
+		log_fatal("Failed to check version compatibility with the monitor "
+				  "extension \"%s\", see above for details",
+				  PG_AUTOCTL_MONITOR_EXTENSION_NAME);
+		return false;
+	}
+
+	if (strcmp(version.installedVersion, PG_AUTOCTL_EXTENSION_VERSION) != 0)
+	{
+		Monitor dbOwnerMonitor = { 0 };
+
+		log_warn("This version of pg_autoctl requires the extension \"%s\" "
+				 "version \"%s\" to be installed on the monitor, current "
+				 "version is \"%s\".",
+				 PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+				 PG_AUTOCTL_EXTENSION_VERSION,
+				 version.installedVersion);
+
+		/*
+		 * Ok, let's try to update the extension then.
+		 *
+		 * For that we need to connect as the owner of the database, which was
+		 * the current $USER at the time of the `pg_autoctl create monitor`
+		 * command.
+		 */
+		if (!prepare_connection_to_current_system_user(monitor,
+													   &dbOwnerMonitor))
+		{
+			log_error("Failed to upgrade extension \"%s\" to version \"%s\": "
+					  "failed prepare a connection string to the "
+					  "monitor as the database owner",
+					  PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+					  PG_AUTOCTL_EXTENSION_VERSION);
+			return false;
+		}
+
+		if (!monitor_extension_update(&dbOwnerMonitor,
+									  PG_AUTOCTL_EXTENSION_VERSION))
+		{
+			log_fatal("Failed to update extension \"%s\" to version \"%s\" "
+					  "on the monitor, see above for details",
+					  PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+					  PG_AUTOCTL_EXTENSION_VERSION);
+			return false;
+		}
+
+		log_info("Updated extension \"%s\" to version \"%s\"",
+				 PG_AUTOCTL_MONITOR_EXTENSION_NAME,
+				 PG_AUTOCTL_EXTENSION_VERSION);
+
+		return true;
+	}
+
+	/* just mention we checked, and it's ok */
+	log_info("The version of extenstion \"%s\" is \"%s\" on the monitor",
+			 PG_AUTOCTL_MONITOR_EXTENSION_NAME, version.installedVersion);
+
+	return true;
+}
+
+
+/*
+ * prepare_connection_to_current_system_user changes a given pguri to remove
+ * its "user" connection parameter, filling in the pre-allocated keywords and
+ * values string arrays.
+ *
+ * Postgres docs at the following address show 30 connection parameters, so the
```

Comment:
Is this perhaps defined in a constant in libpq? would be good if we could use that, otherwise we need to keep track when libpq adds a new parameter.

Reply:
In `postgresql/src/interfaces/libpq/fe-connect.c` we see the following code, which means we're out of luck. I pushed it to 41 (10 extra) and added a check.

``` c
static const internalPQconninfoOption PQconninfoOptions[] = {
...
	/*
	 * Get enough memory for all options in PQconninfoOptions, even if some
	 * end up being filtered out.
	 */
	options = (PQconninfoOption *) malloc(sizeof(PQconninfoOption) * sizeof(PQconninfoOptions) / sizeof(PQconninfoOptions[0]));
```